src/server/pfs/server/driver.go

```diff
@@ -819,57 +819,49 @@ func (d *driver) finishCommit(txnCtx *txnenv.TransactionContext, commit *pfs.Com
 	if description != "" {
 		commitInfo.Description = description
 	}
-	var parentFSID *fileset.ID
+	var ids []fileset.ID
 	if commitInfo.ParentCommit != nil {
 		id, err := d.commitStore.GetTotalFileset(ctx, commitInfo.ParentCommit)
 		if err != nil {
 			return err
 		}
-		parentFSID = id
-	}
-	// Run compaction task.
-	return d.compactionQueue.RunTaskBlock(ctx, func(m *work.Master) error {
-		id, err := d.commitStore.GetDiffFileset(ctx, commit)
-		if err != nil {
-			return err
-		}
-		var ids []fileset.ID
-		// if the commit has a parent, then include the parents fileset in the compaction
-		if parentFSID != nil {
-			ids = append(ids, *parentFSID)
-		}
 		ids = append(ids, *id)
-		compactedID, err := d.compact(m, ids)
-		if err != nil {
-			return err
-		}
-		if err := d.commitStore.SetTotalFileset(ctx, commit, *compactedID); err != nil {
-			return err
-		}
-		outputSize, err := d.storage.SizeOf(ctx, *compactedID)
-		if err != nil {
-			return err
-		}
-		commitInfo.SizeBytes = uint64(outputSize)
-		commitInfo.Finished = types.TimestampNow()
-		empty := strings.Contains(commitInfo.Description, pfs.EmptyStr)
-		if err := d.updateProvenanceProgress(txnCtx, !empty, commitInfo); err != nil {
-			return err
-		}
-		if err := d.writeFinishedCommit(txnCtx.Stm, commit, commitInfo); err != nil {
-			return err
-		}
-		triggeredBranches, err := d.triggerCommit(txnCtx, commitInfo.Commit)
-		if err != nil {
+	}
```

Comment:
So now we are compacting without the lock.  I don't know that it was necessarily the correct lock to be holding, but now it's gone, nonetheless.

We have to make sure no one is adding filesets to the commit when we are compacting. Or at least check that nothing has changed, and call SetTotalFileset in a transaction.

Reply:
Yeah, ideally we would just mark the commit as finished in a transaction (compaction would come later at read time) and that would block later (in the transaction serialization) attempts to modify filesets. I'm not sure what lock you are referring to though. I don't think this was ever locked in a way that would prevent that from occurring. We need to incrementally compact commits anyways to avoid the high fan in of reads on open commits, so I think this would be a reasonable change to include with that change since we would be moving the compaction to read time.

Reply:
Okay I guess mark this as a TODO.  There are a couple of invariants we need to enforce.
1. You cannot write to a closed commit. (probably fixed with Postgres collecitons)
2. If you compact the diff fileset, you cannot write to the total until you check that the diff has not changed in a transaction.